File: thread.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

public class thread {
    // Variabel bersama untuk menghitung jumlah thread yang telah selesai
    private static int completedThreads = 0;
    // Object monitor untuk sinkronisasi
    private static final Object lock = new Object();

    // Thread 1: Menghitung faktorial
    static class FactorialTask implements Runnable {
        private int n;

        public FactorialTask(int n) {
            this.n = n;
        }

        @Override
        public void run() {
            long result = 1;
            for (int i = 1; i <= n; i++) {
                result *= i;
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                }
            }
            System.out.println("Faktorial dari " + n + " = " + result);

            // Sinkronisasi: update counter bersama
            synchronized (lock) {
                completedThreads++;
                System.out.println("Thread faktorial selesai. Total selesai: " + completedThreads);
            }
        }
    }

    // Thread 2: Menampilkan deret fibonacci
    static class FibonacciTask implements Runnable {
        private int n;

        public FibonacciTask(int n) {
            this.n = n;
        }

        @Override
        public void run() {
            int a = 0, b = 1;
            System.out.print("Deret Fibonacci hingga " + n + ": ");
            for (int i = 0; i < n; i++) {
                System.out.print(a + " ");
                int next = a + b;
                a = b;
                b = next;
            }
            System.out.println();

            synchronized (lock) {
                completedThreads++;
                System.out.println("Thread fibonacci selesai. Total selesai: " + completedThreads);
            }
        }
    }

    // Thread 3: Membaca file teks
    static class FileReaderTask implements Runnable {
        private String filename;

        public FileReaderTask(String filename) {
            this.filename = filename;
        }

        @Override
        public void run() {
            try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
                List<String> line = Files.readAllLines(Path.of(filename));
                int index = 1;
                System.out.println("Isi file " + filename + ":");

                for (String b : line) {
                    System.out.println(index + ". " + b);
                    index++;
                    Thread.sleep(100);
                }
            } catch (IOException | InterruptedException e) {
                System.err.println("Error membaca file: " + e.getMessage());
            }

            synchronized (lock) {
                completedThreads++;
                System.out.println("Thread file reader selesai. Total selesai: " + completedThreads);
            }
        }
    }

    public static void main(String[] args) {
        // Buat dan jalankan ketiga thread
        Thread t1 = new Thread(new FactorialTask(5));
        Thread t2 = new Thread(new FibonacciTask(10));
        Thread t3 = new Thread(new FileReaderTask("data.txt"));
        t1.start();
        t2.start();
        t3.start();

        // Tunggu semua thread selesai
        try {
            t1.join();
            t2.join();
            t3.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        System.out.println("Semua thread selesai. Total threads selesai: " + completedThreads);
    }
}

File: thread.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

// Shared variable to count completed threads
static int completed_threads = 0;
// Mutex for synchronization
static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;

// Thread 1: Calculate factorial
typedef struct {
    int n;
} FactorialArgs;

void* factorial_task(void* arg) {
    FactorialArgs* args = (FactorialArgs*)arg;
    int n = args->n;

    long result = 1;
    for (int i = 1; i <= n; i++) {
        result *= i;
        usleep(100000); // 100ms
    }
    printf("Faktorial dari %d = %ld\n", n, result);

    // Synchronize: update shared counter
    pthread_mutex_lock(&lock);
    completed_threads++;
    printf("Thread faktorial selesai. Total selesai: %d\n", completed_threads);
    pthread_mutex_unlock(&lock);

    return NULL;
}

// Thread 2: Display fibonacci sequence
typedef struct {
    int n;
} FibonacciArgs;

void* fibonacci_task(void* arg) {
    FibonacciArgs* args = (FibonacciArgs*)arg;
    int n = args->n;

    int a = 0, b = 1;
    printf("Deret Fibonacci hingga %d: ", n);
    for (int i = 0; i < n; i++) {
        printf("%d ", a);
        int next = a + b;
        a = b;
        b = next;
    }
    printf("\n");

    pthread_mutex_lock(&lock);
    completed_threads++;
    printf("Thread fibonacci selesai. Total selesai: %d\n", completed_threads);
    pthread_mutex_unlock(&lock);

    return NULL;
}

// Thread 3: Read text file
typedef struct {
    const char* filename;
} FileReaderArgs;

void* file_reader_task(void* arg) {
    FileReaderArgs* args = (FileReaderArgs*)arg;
    const char* filename = args->filename;

    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        fprintf(stderr, "Error membaca file: cannot open '%s'\n", filename);

        pthread_mutex_lock(&lock);
        completed_threads++;
        printf("Thread file reader selesai. Total selesai: %d\n", completed_threads);
        pthread_mutex_unlock(&lock);

        return NULL;
    }

    printf("Isi file %s:\n", filename);
    char line[1024];
    int index = 1;
    while (fgets(line, sizeof(line), file)) {
        // Remove trailing newline if present
        int len = 0;
        while (line[len] != '\0') len++;
        if (len > 0 && line[len - 1] == '\n') line[len - 1] = '\0';

        printf("%d. %s\n", index++, line);
        usleep(100000); // 100ms
    }
    fclose(file);

    pthread_mutex_lock(&lock);
    completed_threads++;
    printf("Thread file reader selesai. Total selesai: %d\n", completed_threads);
    pthread_mutex_unlock(&lock);

    return NULL;
}

int main() {
    pthread_t t1, t2, t3;

    // Prepare arguments for each thread
    FactorialArgs  factorial_args  = { .n = 5  };
    FibonacciArgs  fibonacci_args  = { .n = 10 };
    FileReaderArgs file_reader_args = { .filename = "data.txt" };

    // Create and start all three threads
    pthread_create(&t1, NULL, factorial_task,   &factorial_args);
    pthread_create(&t2, NULL, fibonacci_task,   &fibonacci_args);
    pthread_create(&t3, NULL, file_reader_task, &file_reader_args);

    // Wait for all threads to finish
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    pthread_join(t3, NULL);

    printf("Semua thread selesai. Total threads selesai: %d\n", completed_threads);

    pthread_mutex_destroy(&lock);
    return 0;
}